File: python/hard/backtracking_problems.py
"""
Backtracking Problems - N-Queens e Sudoku Solver
"""
from typing import List


def solve_n_queens(n: int) -> List[List[str]]:
    """
    Resolve o problema das N rainhas.
    
    Args:
        n: Tamanho do tabuleiro (n x n)
        
    Returns:
        Lista com todas as soluções possíveis
    """
    result = []
    board = [['.' for _ in range(n)] for _ in range(n)]
    
    def is_valid(row: int, col: int) -> bool:
        for i in range(row):
            if board[i][col] == 'Q':
                return False
        
        i, j = row - 1, col - 1
        while i >= 0 and j >= 0:
            if board[i][j] == 'Q':
                return False
            i -= 1
            j -= 1
        
        i, j = row - 1, col + 1
        while i >= 0 and j < n:
            if board[i][j] = 'Q':
                return False
            i -= 1
            j += 1
        
        return True
    
    def backtrack(row: int):
        if row == n:
            result.append([''.join(r) for r in board])
            return
        
        for col in range(n):
            if is_valid(row, col):
                board[row][col] = 'Q'
                backtrack(row + 1)
                board[row][col] = '.'
    
    backtrack(0)
    return result


def solve_sudoku(board: List[List[str]]) -> None:
    """
    Resolve um Sudoku 9x9 usando backtracking.
    Modifica o board in-place.
    
    Args:
        board: Tabuleiro 9x9 com '.' para células vazias
    """
    def is_valid(row: int, col: int, num: str) -> bool:
        for x in range(9):
            if board[row][x] == num:
                return False
        
        for x in range(9):
            if board[x][col] == num:
                return False
        
        start_row = (row // 3) * 3
        start_col = (col // 3) * 3
        
        for i in range(3):
            for j in range(3):
                if board[start_row + i][start_col + j] == num:
                    return False
        
        return True
    
    def solve() -> bool:
        for row in range(9):
            for col in range(9):
                if board[row][col] == '.':
                    for num in '123456789':
                        if is_valid(row, col, num):
                            board[row][col] = num
                            
                            if solve():
                                return True
                            
                            board[row][col] = '.'
                    return False
        return True
    
    solve()


def permute(nums: List[int]) -> List[List[int]]:
    """
    Gera todas as permutações de uma lista de números.
    """
    result = []
    
    def backtrack(current: List[int], remaining: List[int]):
        if not remaining:
            result.append(current[:])
            return
        
        for i in range(len(remaining)):
            backtrack(current + [remaining[i]], remaining[:i] + remaining[i+1:])
    
    backtrack([], nums)
    return result


if __name__ == "__main__":
    print("N-Queens (n=4):")
    queens_result = solve_n_queens(4)
    for idx, solution in enumerate(queens_result):
        print(f"Solução {idx + 1}:")
        for row in solution:
            print(row)
        print()
    
    print("\nSudoku Solver:")
    sudoku_board = [
        ["5","3",".",".","7",".",".",".","."],
        ["6",".",".","1","9","5",".",".","."],
        [".","9","8",".",".",".",".","6","."],
        ["8",".",".",".","6",".",".",".","3"],
        ["4",".",".","8",".","3",".",".","1"],
        ["7",".",".",".","2",".",".",".","6"],
        [".","6",".",".",".",".","2","8","."],
        [".",".",".","4","1","9",".",".","5"],
        [".",".",".",".","8",".",".","7","9"]
    ]
    solve_sudoku(sudoku_board)
    print("Sudoku resolvido:")
    for row in sudoku_board
        print(' '.join(row))
    
    print("\nPermutations:")
    print(permute([1, 2, 3]))

File: javascript/hard/backtracking_problems.js
/**
 * N-Queens Problem
 * Coloca N rainhas em um tabuleiro NxN sem que se ataquem
 * @param {number} n
 * @return {string[][]}
 */
function solveNQueens(n) {
    const result = [];
    const board = Array(n).fill().map(() => Array(n).fill('.'));
    
    function isValid(row, col) {
        for (let i = 0; i < row; i++) {
            if (board[i][col] === 'Q') {
                return false;
            }
        }
        
        for (let i = row - 1, j = col - 1; i >= 0 && j >= 0; i--, j--) {
            if (board[i][j] === 'Q') {
                return false;
            }
        }
        
        for (let i = row - 1, j = col + 1; i >= 0 && j < n; i--, j++) {
            if (board[i][j] = 'Q') {
                return false;
            }
        }
        
        return true;
    }
    
    function backtrack(row) {
        if (row === n) {
            result.push(board.map(r => r.join('')));
            return;
        }
        
        for (let col = 0; col < n; col++) {
            if (isValid(row, col)) {
                board[row][col] = 'Q';
                backtrack(row + 1);
                board[row][col] = '.';
            }
        }
    }
    
    backtrack(0);
    return result;
}

/**
 * Sudoku Solver
 * Resolve um Sudoku usando backtracking
 * @param {character[][]} board
 * @return {void}
 */
function solveSudoku(board) {
    function isValid(row, col, num) {
        for (let x = 0; x < 9; x++) {
            if (board[row][x] === num) {
                return false;
            }
        }
        
        for (let x = 0; x < 9; x++) {
            if (board[x][col] === num) {
                return false;
            }
        }
        
        const startRow = Math.floor(row / 3) * 3;
        const startCol = Math.floor(col / 3) * 3;
        
        for (let i = 0; i < 3; i++) {
            for (let j = 0; j < 3; j++) {
                if (board[startRow + i][startCol + j] === num) {
                    return false;
                }
            }
        }
        
        return true;
    }
    
    function solve() {
        for (let row = 0; row < 9; row++) {
            for (let col = 0; col < 9; col++) {
                if (board[row][col] === '.') {
                    for (let num = '1'; num <= '9'; num++) {
                        if (isValid(row, col, num)) {
                            board[row][col] = num;
                            
                            if (solve()) {
                                return true;
                            }
                            
                            board[row][col] = '.';
                        }
                    }
                    return false;
                }
            }
        }
        return true;
    }
    
    solve();
}

console.log("N-Queens (n=4):");
const queensResult = solveNQueens(4);
queensResult.forEach((solution, idx) => {
    console.log(`Solução ${idx + 1}:`);
    solution.forEach(row => console.log(row));
    console.log();
});

console.log("\nSudoku Solver:");
const sudokuBoard = [
    ["5","3",".",".","7",".",".",".","."],
    ["6",".",".","1","9","5",".",".","."],
    [".","9","8",".",".",".",".","6","."],
    ["8",".",".",".","6",".",".",".","3"],
    ["4",".",".","8",".","3",".",".","1"],
    ["7",".",".",".","2",".",".",".","6"],
    [".","6",".",".",".",".","2","8","."],
    [".",".",".","4","1","9",".",".","5"],
    [".",".",".",".","8",".",".","7","9"]
];
solveSudoku(sudokuBoard);
console.log("Sudoku resolvido:")
sudokuBoard.forEach(row => console.log(row.join(' ')));
